File: scripts/test_tabs.py
import time
import sys
import subprocess
import signal
import os
from playwright.sync_api import sync_playwright

# Tab pages to test
TABS = [
    {"path": "/home/index.html", "name": "Home", "title": "Khyati"},
    {"path": "/about/about.html", "name": "About", "title": "About"},
    {"path": "/ai-tools/ai-tools.html", "name": "AI Tools", "title": "AI Tools"},
    {"path": "/fun-games/fun-games.html", "name": "Fun & Games", "title": "Fun"},
    {"path": "/trading/trading.html", "name": "Trading", "title": "Trading"},
    {"path": "/stories/stories.html", "name": "Stories", "title": "Stories"},
    {"path": "/videos/videos.html", "name": "Videos", "title": "Videos"},
    {"path": "/others/others.html", "name": "Others", "title": "Others"},
    {"path": "/admin/login.html", "name": "Login", "title": "Sign in"},
    {"path": "/admin/signup.html", "name": "Sign Up", "title": "Sign up"},
    {"path": "/index.html", "name": "Redirect", "expected_redirect": "/home/index.html"},
]


def start_server(project_dir):
    """Start a simple HTTP server for the project."""
    proc = subprocess.Popen(
        [sys.executable, "-m", "http.server", "8000", "--directory", project_dir],
        stdout=subprocess.DEVNULL,
        stderr=subprocess.DEVNULL,
    )
    time.sleep(1)
    return proc


def run_tests(project_dir):
    server = start_server(project_dir)
    errors = []
    try:
        with sync_playwright() as p:
            browser = p.chromium.launch(headless=True)
            context = browser.new_context(
                viewport={"width": 1280, "height": 800},
                ignore_https_errors=True,
            )
            page = context.new_page()

            for tab in TABS:
                url = f"http://localhost:8000{tab['path']}"
                print(f"\n{'='*60}")
                print(f"Testing: {tab['name']} -> {url}")
                try:
                    page.goto(url, wait_until="networkidle", timeout=15000)

                    # Check for redirect
                    if "expected_redirect" in tab:
                        time.sleep(0.5)
                        current = page.url
                        if tab["expected_redirect"] in current:
                            print(f"  ✓ Redirect working: {current}")
                        else:
                            msg = f"  ✗ Expected redirect to {tab['expected_redirect']}, got {current}"
                            print(msg)
                            errors.append(msg)
                        continue

                    # Check page loaded (no crash, body present)
                    body = page.query_selector("body")
                    if not body:
                        msg = f"  ✗ No body element found"
                        print(msg)
                        errors.append(msg)
                        continue

                    # Check title contains expected text
                    actual_title = page.title()
                    if tab["title"].lower() in actual_title.lower():
                        print(f"  ✓ Title matches: '{actual_title}'")
                    else:
                        msg = f"  ✗ Expected title containing '{tab['title']}', got '{actual_title}'"
                        print(msg)
                        errors.append(msg)

                    # Check header is loaded (site-header class)
                    header = page.query_selector(".site-header")
                    if header:
                        print(f"  ✓ Header loaded")
                    else:
                        print(f"  ⚠ Header not found (may be loaded via JS)")

                    # Check footer is present
                    footer = page.query_selector(".site-footer")
                    if footer:
                        print(f"  ✓ Footer loaded")
                    else:
                        print(f"  ⚠ Footer not found")

                    # Check main content area
                    main = page.query_selector("main")
                    if main:
                        print(f"  ✓ Main content area found")
                    else:
                        print(f"  ⚠ Main element not found")

                    print(f"  ✓ OK - Page loaded successfully")

                except Exception as e:
                    msg = f"  ✗ FAILED: {tab['name']} - {str(e)}"
                    print(msg)
                    errors.append(msg)

            browser.close()

        print(f"\n{'='*60}")
        if errors:
            print(f"\n❌ {len(errors)} test(s) failed:")
            for e in errors:
                print(f"  - {e}")
            print(f"\n❌ Some tests FAILED")
            sys.exit(1)
        else:
            print(f"\n✅ ALL {len(TABS)} tests PASSED!")
            sys.exit(0)

    finally:
        server.terminate()
        try:
            server.wait(timeout=5)
        except:
            server.kill()


if __name__ == "__main__":
    project_dir = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
    print(f"Project directory: {project_dir}")
    run_tests(project_dir)

File: scripts/test_tabs.mjs
import { chromium } from 'playwright';
import { createServer } from 'http';
import { readFileSync, existsSync, statSync } from 'fs';
import { join, extname } from 'path';
import { fileURLToPath } from 'url';
import { platform, homedir } from 'os';

const __dirname = join(fileURLToPath(import.meta.url), '..');
const projectDir = join(__dirname, '..');

const MIME_TYPES = {
  '.html': 'text/html',
  '.js': 'application/javascript',
  '.mjs': 'application/javascript',
  '.css': 'text/css',
  '.json': 'application/json',
  '.png': 'image/png',
  '.jpg': 'image/jpeg',
  '.jpeg': 'image/jpeg',
  '.gif': 'image/gif',
  '.svg': 'image/svg+xml',
  '.ico': 'image/x-icon',
  '.woff': 'font/woff',
  '.woff2': 'font/woff2',
  '.txt': 'text/plain',
  '.md': 'text/plain',
};

const TABS = [
  { path: '/home/index.html', name: 'Home', title: 'Khyati' },
  { path: '/about/about.html', name: 'About', title: 'About' },
  { path: '/ai-tools/ai-tools.html', name: 'AI Tools', title: 'AI Tools' },
  { path: '/ai-tools/news-app.html', name: 'News App', title: 'News' },
  { path: '/ai-tools/voice-clone.html', name: 'Voice Clone', title: 'Voice' },
  { path: '/ai-tools/story-creator.html', name: 'Story Creator', title: 'Story Creator' },
  { path: '/fun-games/fun-games.html', name: 'Fun & Games', title: 'Fun' },
  { path: '/trading/trading.html', name: 'Trading', title: 'Trading' },
  { path: '/stories/stories.html', name: 'Stories', title: 'Stories' },
  { path: '/videos/videos.html', name: 'Videos', title: 'Videos' },
  { path: '/others/others.html', name: 'Others', title: 'Others' },
  { path: '/admin/login.html', name: 'Login', title: 'Sign in' },
  { path: '/admin/signup.html', name: 'Sign Up', title: 'Sign up' },
  { path: '/index.html', name: 'Root (canonical homepage)', title: 'Khyati' },
];

function startServer(port) {
  return new Promise((resolve) => {
    const server = createServer((req, res) => {
      let filePath = join(projectDir, req.url === '/' ? '/index.html' : req.url);
      // Decode URL
      filePath = decodeURIComponent(filePath);

      if (!existsSync(filePath)) {
        res.writeHead(404, { 'Content-Type': 'text/plain' });
        res.end('Not Found');
        return;
      }

      const ext = extname(filePath).toLowerCase();
      const contentType = MIME_TYPES[ext] || 'application/octet-stream';
      const content = readFileSync(filePath);
      res.writeHead(200, { 'Content-Type': contentType });
      res.end(content);
    });

    server.listen(port, () => {
      console.log(`Test server running on http://localhost:${port}`);
      resolve(server);
    });
  });
}

async function runTests() {
  const PORT = 8000;
  const server = await startServer(PORT);
  let exitCode = 0;
  const errors = [];

  try {
    const browser = await chromium.launch({ headless: true });
    const context = await browser.newContext({
      viewport: { width: 1280, height: 800 },
    });
    const page = await context.newPage();

    for (const tab of TABS) {
      const url = `http://localhost:${PORT}${tab.path}`;
      console.log(`\n${'='.repeat(60)}`);
      console.log(`Testing: ${tab.name} -> ${url}`);

      try {
        await page.goto(url, { waitUntil: 'networkidle', timeout: 15000 });

        // Handle redirect test
        if (tab.expectedRedirect) {
          await page.waitForTimeout(500);
          const currentUrl = page.url();
          if (currentUrl.includes(tab.expectedRedirect)) {
            console.log(`  ✓ Redirect working: ${currentUrl}`);
          } else {
            const msg = `  ✗ Expected redirect to ${tab.expectedRedirect}, got ${currentUrl}`;
            console.log(msg);
            errors.push(msg);
            exitCode = 1;
          }
          continue;
        }

        // Check body exists
        const body = await page.$('body');
        if (!body) {
          const msg = `  ✗ No body element found`;
          console.log(msg);
          errors.push(msg);
          exitCode = 1;
          continue;
        }

        // Check title
        const actualTitle = await page.title();
        if (actualTitle.toLowerCase().includes(tab.title.toLowerCase())) {
          console.log(`  ✓ Title: '${actualTitle}'`);
        } else {
          const msg = `  ✗ Expected title containing '${tab.title}', got '${actualTitle}'`;
          console.log(msg);
          errors.push(msg);
          exitCode = 1;
        }

        // Check header
        const header = await page.$('.site-header, #site-header-root');
        if (header) {
          console.log(`  ✓ Header element found`);
        } else {
          console.log(`  ⚠ Header not found`);
        }

        // Check footer
        const footer = await page.$('.site-footer');
        if (footer) {
          console.log(`  ✓ Footer found`);
        } else {
          console.log(`  ⚠ Footer not found`);
        }

        // Check main content
        const main = await page.$('main, .hero');
        if (main) {
          console.log(`  ✓ Main content area found`);
        } else {
          console.log(`  ⚠ Main content not found`);
        }

        // Check for JS errors
        const jsErrors = [];
        page.on('pageerror', err => jsErrors.push(err.message));
        if (jsErrors.length > 0) {
          console.log(`  ⚠ JS Errors: ${jsErrors.join(', ')}`);
        }

        console.log(`  ✓ PASSED: ${tab.name}`);
      } catch (err) {
        const msg = `  ✗ FAILED: ${tab.name} - ${err.message}`;
        console.log(msg);
        errors.push(msg);
        exitCode = 1;
      }
    }

    await browser.close();
  } finally {
    server.close();
  }

  console.log(`\n${'='.repeat(60)}`);
  if (errors.length > 0) {
    console.log(`\n❌ ${errors.length} test(s) failed:`);
    errors.forEach(e => console.log(`  - ${e}`));
    console.log(`\n❌ Some tests FAILED`);
  } else {
    console.log(`\n✅ ALL ${TABS.length} tests PASSED!`);
  }

  process.exit(exitCode);
}

runTests().catch(err => {
  console.error('Fatal error:', err);
  process.exit(1);
});